http/basic/src/main/java/io/cloudevents/http/HttpMessageFactory.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudevents.http;
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpsExchange;
+import io.cloudevents.core.message.MessageReader;
+import io.cloudevents.core.message.MessageWriter;
+import io.cloudevents.core.message.impl.GenericStructuredMessageReader;
+import io.cloudevents.core.message.impl.MessageUtils;
+import io.cloudevents.core.message.impl.UnknownEncodingMessageReader;
+import io.cloudevents.http.impl.CloudEventsHeaders;
+import io.cloudevents.http.impl.HttpMessageReader;
+import io.cloudevents.http.impl.HttpMessageWriter;
+
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URLConnection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import static io.cloudevents.http.impl.CloudEventsHeaders.CONTENT_TYPE;
+
+/**
+ * This class provides a collection of methods to create {@link io.cloudevents.core.message.MessageReader}
+ * and {@link io.cloudevents.core.message.MessageWriter} for various HTTP APIs.
+ */
+public final class HttpMessageFactory {
+
+    private HttpMessageFactory() {}
+
+    /**
+     * Creates a new {@link MessageReader} that can read both structured and binary messages from a HTTP response (client) or request (server).
+     *
+     * <pre>
+     * Example of usage with <a href="https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html">HttpServletRequest</a>:
+     * {@code
+     * Consumer<BiConsumer<String,String>> forEachHeader = processHeader -> {
+     *     Enumeration<String> headerNames = httpServletRequest.getHeaderNames();
```

Comment:
```
for (String name : Collections.list(httpServletRequest.getHeaderNames())) {
```

Reply:
Definitely would work, but why allocate new List just to iterate over it?

Reply:
`Collections.list` was introduced specifically to get the legacy type `Enumeration` out of the way as quickly as possible. For me this is the idiomatic way to iterate over the `Enumeration<String>` type that the ancient Servlet API imposes. It's also a lot more readable for the example code.

Reply:
Just found out that there is `asIterator()` then I could call `forEachRemaining` unfortunately the `asIterator()` method is available since Java 9+ :disappointed:

Reply:
I agree that
```java
for (String name : Collections.list(httpServletRequest.getHeaderNames())) {}
```
is probably more readable.

But I find `ArrayList` allocation (and possible internal reallocation on its growth) unnecessary.
I specifically chosen that obsolete type to show that it works well with `Consumer<BiConsumer<String,String>> forEachHeader`. @slinkydeveloper wdyt?